### PR TITLE
[stable/rabbitmq] Add Prometheus operator ServiceMonitor

### PR DESCRIPTION
**Pull request by moebius87**
#### What this PR does / why we need it:
Simplify Prometheus integration by creating ServiceMonitor Resource for scraping metrics using PrometheusOperator

#### Which issue this PR fixes
None.

#### Special notes for your reviewer:
None.

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/chart]`)

### CODE REVIEW

File: stable/rabbitmq/templates/servicemonitor.yaml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    chart: {{ template "rabbitmq.chart" .  }}
+    heritage: "{{ .Release.Service }}"
+    release: "{{ .Release.Name }}"
+    {{- if .Values.metrics.serviceMonitor.additionalLabels }}
+{{ toYaml .Values.metrics.serviceMonitor.additionalLabels | indent 4 }}
+    {{- end }}
+spec:
+  endpoints:
+  - port: metrics
+    interval: 30s
```
Comment (juan131) [inline stable/rabbitmq/templates/servicemonitor.yaml]: The interval should be configurable. See PostgreSQL for instance:
https://github.com/helm/charts/blob/master/stable/postgresql/templates/servicemonitor.yaml#L21

Comment (juan131) [reply stable/rabbitmq/templates/servicemonitor.yaml]: The `scrapeTimeout` should be also configurable. See PostgreSQL for instance:
https://github.com/helm/charts/blob/master/stable/postgresql/templates/servicemonitor.yaml#L24

File: stable/rabbitmq/templates/servicemonitor.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    {{- if .Values.metrics.serviceMonitor.honorLabels }}
+    honorLabels: true
+    {{- end }}
+    {{- if .Values.metrics.serviceMonitor.relabellings }}
+    metricRelabelings:
+{{ toYaml .Values.metrics.serviceMonitor.relabellings | indent 6 }}
+    {{- end }}
+  namespaceSelector:
+    matchNames:
+      - {{ .Release.Namespace }}
```
Comment (juan131) [inline stable/rabbitmq/templates/servicemonitor.yaml:31]: Shouldn't you use `{{ .Values.metrics.serviceMonitor.namespace }}` here?

File: stable/rabbitmq/values-production.yaml [hunk 3]
```diff
@@ -293,6 +293,25 @@ metrics:
     ##
     # pullSecrets:
     #   - myRegistryKeySecretName
+    
+  ## Prometheus Service Monitor
+  ## ref: https://github.com/coreos/prometheus-operator
+  ##      https://github.com/coreos/prometheus-operator/blob/master/Documentation/api.md
+  serviceMonitor:
+    ## If the operator is installed in your cluster, set to true to create a Service Monitor Entry
+    enabled: true
```
Comment (juan131) [inline stable/rabbitmq/values-production.yaml]: Do not enable it by default in the **values-production.yaml**.
ServiceMonitor is a CRD and it won't be available in the API unless the Prometheus Operator is installed in the cluster. Therefore, the user must implicitly choose to enable it.

**Review by juan131 — CHANGES_REQUESTED** [review]
Thanks so much for the contribution @moebius87 
Please review my comments when you have a chance

**Comment by tompizmor** [comment]
@moebius87 Once you address Juan suggestions, can you please also take care of solving the conflicts in `Chart.yaml` and `values.yaml`?

File: stable/rabbitmq/templates/servicemonitor.yaml (hunk 1 above)
Comment (moebius87) [reply stable/rabbitmq/templates/servicemonitor.yaml]: Done.

File: stable/rabbitmq/templates/servicemonitor.yaml (hunk 2 above)
Comment (moebius87) [reply stable/rabbitmq/templates/servicemonitor.yaml:31]: No, is correct right that. That parameter must match the namespace of [rabbimq service resource](https://github.com/helm/charts/blob/master/stable/rabbitmq/templates/svc.yaml).

See these other charts for check:
* https://github.com/helm/charts/blob/master/stable/postgresql/templates/servicemonitor.yaml#L28
* https://github.com/helm/charts/blob/master/stable/mongodb/templates/prometheus-service-monitor.yaml#L25
* https://github.com/helm/charts/blob/master/stable/redis/templates/metrics-prometheus.yaml#L29

File: stable/rabbitmq/values-production.yaml (hunk 3 above)
Comment (moebius87) [reply stable/rabbitmq/values-production.yaml]: It seems to me that, if a CRD is not available in the API, helm not returns errors; it simply not create the resource.
However, you are right and I followed your advice. Now the user must explicitly choose to enable it.

**Comment by moebius87** [comment]
@juan131 @tompizmor All comments are resolved.
( Thank you for your works! 🥇 )

File: stable/rabbitmq/values-production.yaml [hunk 4]
```diff
@@ -292,6 +292,7 @@ metrics:
     ##
     # pullSecrets:
     #   - myRegistryKeySecretName
+
```
Comment (tompizmor) [inline stable/rabbitmq/values-production.yaml]: Please either remove this blank line or also add it to the values-production.yaml

File: stable/rabbitmq/README.md [hunk 5]
```diff
@@ -119,6 +119,13 @@ The following table lists the configurable parameters of the RabbitMQ chart and
 | `metrics.image.repository`           | Exporter image name                              | `kbudde/rabbitmq-exporter`                              |
 | `metrics.image.tag`                  | Exporter image tag                               | `v0.29.0`                                               |
 | `metrics.image.pullPolicy`           | Exporter image pull policy                       | `IfNotPresent`                                          |
+| `metrics.serviceMonitor.enabled`     | Create ServiceMonitor Resource for scraping metrics using PrometheusOperator   | `false`                   |
+| `metrics.serviceMonitor.namespace`   | Namespace where servicemonitor resource should be created         | `the same namespace as nginx ingress`  |
```
Comment (tompizmor) [inline stable/rabbitmq/README.md]: `the same namespace as nginx ingress` is this default value correct? In the values files I see the namespace not set so you should use `nil`

**Review by tompizmor — CHANGES_REQUESTED** [review]

**Comment by tompizmor** [comment]
@moebius87 I just added a couple of comments in the PR! Please take a look when you have some time

File: stable/rabbitmq/README.md (hunk 5 above)
Comment (moebius87) [reply stable/rabbitmq/README.md]: Sorry, distraction error. You're absolutely right. I fixed it.

File: stable/rabbitmq/values-production.yaml (hunk 4 above)
Comment (moebius87) [reply stable/rabbitmq/values-production.yaml]: they are now realigned

**Review by juan131 — APPROVED** [review]
Thanks so much for addressing all the requested changes! I'm approving your PR

**Pull request merged**